### DISCUSSION TITLE
Pruning with multi-objective study

### DISCUSSION THREAD

**Discussion by ConnorBaker** [optuna/optuna#4107] [Q&A]
Hi all,

I've got a project I'm working on (https://github.com/connorbaker/bsrt) and would like to get started doing hyperparameter optimization.

As a super-resolution project, while some metrics are better than others (LPIPS vs MSE, for example), there isn't any one metric which accurately captures everything (or even most of the things) we care about. As such, I've made a multi-objective study using LPIPS, MS-SSIM, and PSNR because these are three of the most important metrics I want to optimize for. While LPIPS is a rough proxy for the values of MS-SSIM and PSNR, there is still some variance, so it's important that I have these additional metrics to use to order trial results. Unfortunately, it takes quite a while to train the model, I need something like early stopping to better explore the parameter space.

I know that with $n$-objective studies we have an $n$-dimensional Pareto front (Pareto surface?). Is there any way to prune trials in a multi-objective study which are likely to yield a point on the interior of the Pareto front? If not, is there a way to prune trials based on only one objective of a multi-objective study? In my example, I'd like to prune values based on all three metrics, but would settle for just pruning based on LPIPS because that should be enough to get to parameters which yield good MS-SSIM and PSNR values.

I look forward from hearing from you all. Thank you for all of the work you've put into Optuna and for making it open source!

Warm regards,
Connor Baker

**Comment by nzw0301** [comment]
Unfortunately, optuna doesn't support its Pruner class with multi-objective study as described in 
 https://github.com/optuna/optuna/issues/3450. I'm not sure what happens when we store intermediate values as user attributes and manically raise `optuna.TrialPruned()` in multi-objective study...

**Reply by ConnorBaker** [reply]
I made a fork of Optuna (https://github.com/ConnorBaker/optuna/tree/feature/multi-objective-refactor) where I’ve been trying to implement it. The main changes I’ve been working on are:

1. `value` is now a `TrialValue`, an ordered dictionary mapping metric name to observed value. In the single-objective case, `value` has one metric.
2. `directions` is now a `StudyDirections`, an ordered dictionary mapping metric name to `StudyDirection`. In the single-objective case, `directions` has one metric. 
3. Removing the `values` and `direction` attributes and instead using `value` and `directions` (although… I should think again about what should be plural and what shouldn’t be — we already have a `StudyDirection` enum and I didn’t want to replace that which is why `directions` and `StudyDirections` are plural)
4. Adding functionality to the pruners so each can declare whether or not they support multi-objective pruning. In the case that they don’t and there’s more than one key in the reported `TrialValue`, the pruner should be initialized with a `prune_on_metric: MetricName` attribute which instructs it to ignore the other metrics and function as if it were run on a single-objective study.
5. Updating the storage classes to use `dataclass` where possible instead of passing around untyped dictionaries.

Is this at all interesting? I realize there are LOTS of backwards-incompatible changes and have mostly been doing this to see what would happen, not with the intent of contributing back upstream —- I would certainly try to discuss that with you all before doing that.

**Reply by nzw0301** [reply]
Awesome!!

Let me ping a few core members who participated in the related issue @HideakiImamura @g-votte because I don't have time to discuss this feature soon...
